### PR TITLE
Ensure Parameter re-validates default if slot changes

### PR DESCRIPTION
**Pull request by philippjfr**
This PR adds validation when overriding slot values on a Parameter in a subclass. It does this by checking if a superclass defines a slot value and if that slot value is overridden by the subclass. It attempts to minimize the number of actual lookups by aborting early if we already know that re-validation has to happen and by skipping checks on various slot values which do not affect validation.

Fixes https://github.com/holoviz/param/issues/714

### CODE REVIEW

**Comment by philippjfr** [comment]
I had to disable re-validation of None values because the automatic appending of an unknown value on `Selector` opens a whole rabbit hole in regard to the validation.

File: param/parameterized.py [hunk 1]
```diff
@@ -3325,27 +3332,35 @@ def __param_inheritance(mcs, param_name, param):
...
+                new_value = getattr(new_param, slot)
+                old_value = slot_values.get(slot, Undefined)
+                if new_value is Undefined or new_value is old_value:
+                    continue
+                elif old_value is Undefined:
+                    slot_values[slot] = new_value
+                    # If we already know we have to re-validate abort
+                    # early to avoid costly lookups
+                    if slot_overridden or type_change:
+                        continue
```
Comment (maximlt) [inline param/parameterized.py:3357]: As far as I can see this `continue` is not skipping any code path, was it supposed to be a `break`?

**Review by jbednar — COMMENTED** [review]
I'm having a good bit of trouble following the logic here, so hopefully we can work through it together. I think I probably wrote some of the confusing bits (e.g. what's "new" about what I think I was the one to call "new_param"?), but as the code has gotten more complex I'm now lost. The basic idea seems sound, i.e. to see if any slots have changed and thus whether we need to re-validate the default value.

File: param/parameterized.py (hunk 1 above)
Comment (philippjfr) [inline param/parameterized.py:3357]:
```suggestion
                        break
```

**Comment by philippjfr** [comment]
@maximlt Can you either merge or ask for further clarification? You are release manager :)

**Comment by maximlt** [comment]
I was waiting for https://github.com/holoviz/param/pull/826 to be merged! Now it's done I'll rebase this and merge if it goes well.

**Pull request merged**